### PR TITLE
Fix gradient checkpointing imagegpt

### PR DESCRIPTION
**Pull request by KMFODA**
This PR fixes a bug that a user can encounter while using generate and models that use gradient_checkpointing.

Fixes Issue https://github.com/huggingface/transformers/issues/21737

cc @younesbelkada or @gante

### CODE REVIEW

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 1]
```diff
@@ -500,6 +500,13 @@ def forward(
 
         hidden_states = inputs_embeds
 
+        if self.gradient_checkpointing and self.training:
+            if use_cache:
+                logger.warning(
+                    "`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`..."
+                )
+                use_cache = False
+
```
Comment (younesbelkada) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:509]:
```suggestion
```

**Review by younesbelkada — COMMENTED** [review]
Thanks for your work on this! 🔥 
Left a tiny suggestion, otherwise LGTM!

**Review by younesbelkada — APPROVED** [review]
Thanks for your great work on this! 💯

**Review by gante — APPROVED** [review]
Thank you for the contribution 🙏
